src/Microsoft.Health.Api/Features/HealthChecks/CachedHealthCheck.cs

```diff
@@ -10,91 +10,133 @@
+        private volatile CachedHealthCheckResult _cachedResult = new CachedHealthCheckResult();
 
-            _provider = provider;
-            _healthCheck = healthCheck;
-            _logger = logger;
+        public CachedHealthCheck(IServiceProvider provider, Func<IServiceProvider, IHealthCheck> healthCheck)
+        {
+            _provider = EnsureArg.IsNotNull(provider, nameof(provider));
+            _healthCheck = EnsureArg.IsNotNull(healthCheck, nameof(healthCheck));
+            _options = provider.GetRequiredService<IOptions<HealthCheckCachingOptions>>()?.Value;
+            _logger = provider.GetRequiredService<ILogger<CachedHealthCheck>>();
         }
 
-        private static DateTimeOffset ExpirationWindow => Clock.UtcNow.Add(-CacheTime);
-
-        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
         {
-            if (_lastChecked >= ExpirationWindow)
+            DateTimeOffset currentTime = Clock.UtcNow;
+            if (IsCacheFresh(currentTime))
             {
-                return _lastResult;
+                return Task.FromResult(_cachedResult.Value);
             }
 
+            return RefreshCache(context, currentTime, cancellationToken);
+        }
+
+        private async Task<HealthCheckResult> RefreshCache(HealthCheckContext context, DateTimeOffset currentTime, CancellationToken cancellationToken)
+        {
+            // If the cache is stale but not expired, then we make a best effort to refresh if the semaphore isn't
+            // currently occupied by another thread. Otherwise, we'll use the cached value.
+            TimeSpan maxWait = currentTime < _cachedResult.ExpireTime ? TimeSpan.Zero : Timeout.InfiniteTimeSpan;
+
             try
             {
-                await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                // Note: Only the timespan controls whether or not the method returns true/false.
+                //       If the token is canceled, the method throws instead an OperationCanceledException.
+                if (!await _semaphore.WaitAsync(maxWait, cancellationToken).ConfigureAwait(false))
+                {
+                    return _cachedResult.Value;
+                }
             }
-            catch (OperationCanceledException oce) when (cancellationToken.IsCancellationRequested)
+            catch (OperationCanceledException oce) when (oce.CancellationToken == cancellationToken)
             {
-                _logger.LogInformation(oce, $"Cancellation was requested for {nameof(CheckHealthAsync)}");
-                return _lastResult;
+                _logger.LogWarning(oce, $"Cancellation was requested for {nameof(CheckHealthAsync)}.");
+                throw;
```

Comment:
I guess it doesn't matter because the operation is cancelled, but what response does this actually return?

Reply:
Would it make sense to return the existing status if we're going to throw it away anyway?

Reply:
It returns "Unhealthy". In the case of a user's cancellation, they'll never see the response since they aborted the connection. But if we timeout (and the result has expired), we'll throw. Perhaps I can change the catch block later to also check for a fresh result.

Reply:
Updated